torch/_subclasses/fake_tensor.py

```diff
@@ -51,6 +52,90 @@ def _is_tensor_constructor(func: OpOverload):
 # FakeTensor extends MetaTensors to also carry an additional `fake_device`
 # which tracks devices that would have been used.
 
+def torch_dispatch_impl(cls_or_mode_instance, func, types, args, kwargs, run_function):
+    kwargs = kwargs if kwargs else {}
+
+    # This classes virtualizes .device() calls, need to short-circuit
+    # it insteead of calling device again or we would keep on recurring
+    if func == torch.ops.prim.device.default:
+        assert len(args) == 1 and isinstance(args[0], FakeTensor)
+        return args[0].fake_device
+
+    def wrap(e, device=None):
+        if isinstance(e, torch.Tensor) and not isinstance(e, FakeTensor):
+            if device:
+                return FakeTensor(e, device)
+            else:
+                return FakeTensor.from_tensor(e)
+        else:
+            return e
+
+    # if we are in the dispatch mode, we will enter this function even if the inputs
+    # are not FakeTensors. For now, throw if any non-Fake Tensor inputs
+    # and just support constructors. TODO: extend more broadly
+    if isinstance(cls_or_mode_instance, FakeTensorMode):
+        conversion_made = False
+
+        def check_non_fake_tensor(x):
+            nonlocal conversion_made
+            conversion_made = conversion_made or (isinstance(x, torch.Tensor) and not isinstance(x, FakeTensor))
+
+        tree_map(check_non_fake_tensor, args)
+        tree_map(check_non_fake_tensor, kwargs)
+
+        if conversion_made:
+            raise Exception(
+                "Invoking operators with non-Fake Tensor inputs in FakeTensorMode is not yet supported. "
+                f"Please convert all Tensors to FakeTensors first. Found in {func}"
+            )
+
+    # _to_copy fails when run with FakeTensors to cuda device
+    # TODO: debug
+    if func == torch.ops.aten._to_copy.default:
+        _, new_kwargs = normalize_function(
+            func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+        )
+        out_device = new_kwargs.pop("device", new_kwargs["input"].device)
+        with no_dispatch():
+            input = new_kwargs.pop("input").to("meta")
+            return FakeTensor(
+                torch.ops.aten._to_copy(input, **new_kwargs), out_device
+            )
+
+    if _is_tensor_constructor(func):
+        assert func not in _non_kwarg_device_constructors
+        _, new_kwargs = normalize_function(
+            func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+        )
+        # cpu is default device if none is specified
+        out_device = new_kwargs.pop("device", torch.device("cpu"))
+        new_kwargs["device"] = torch.device("meta")
+        r = run_function(func, types, (), new_kwargs)
+        return FakeTensor(r, out_device)
```

Comment:
should be in `no_dispatch` to allow for nested modes (more info https://github.com/pytorch/pytorch/pull/78516/files#r885905994)